Partial solution to #18

JamesSaxon opened the pull request:
This solves the issue noted by @knaaptime in #18, in his notebook, that the two-stage numbers were not affected by changes in the OD matrix.  (Thanks!)

This fixes a bug where it was not the weighted values that were getting summed in the two-stage calculation!  This was basically a one-line bug, that the suffix "r" [had been added](https://github.com/pysal/access/blob/bfae8edbbcb304e67884a3b1cbf292cc410fa380/access/fca.py#L224), but then not used.

I wrote a completely "base python" gravity-based 2SFCA below, that we could add as an additional unit test.'  With that fix, the package and the function below (for the four simple cases!!) agree, point-identical.  

https://gist.github.com/4a3c3d1c571523db43785e43b256ba78

See if this looks reasonable?

```
def simple_2sfca(OD, supply = docs, demand = pops, locs = locs, max_travel = 61,
                 return_debug = False):
    
    D = {hosp : sum(demand[res] / OD[res][hosp]
                    for res in locs 
                    if OD[res][hosp] < max_travel)
         for hosp in locs}

    R = {l : supply[l] / D[l] for l in locs}
    
    A = {res : sum(R[hosp] / OD[res][hosp] 
                   for hosp in locs
                    if OD[res][hosp] < max_travel)
         for res in locs}
    
    if return_debug:
        return A, R, D
    
    return A
```

knaaptime commented:
nice. I dont know that I ever would've caught that. 

2 quick things:
- Its not critical, but are you cool with making the threshold <=? I think its more commomn for people to think of the max cost as something they're willing to spend up-to but not over.
- do you want to add someone to [co-maintain](https://docs.google.com/spreadsheets/d/186MdVDgPpqb_PBPfbEk32f7kW8L4V_ukxkuBhrvL5r8/edit#gid=0)? I'm happy to do it unless you'd like to propose someone better suited

JamesSaxon commented:
Great, can clean up once I have time to dig into the failing tests.

Yes, totally fine with switching to <=, though I think that's logically another issue.

@jkoschinsky is great if she's willing!!

jkoschinsky commented:
Yes, great, thanks!!
________________________________
From: Jamie Saxon ***@***.***>
Sent: Friday, February 4, 2022 10:00:10 PM
To: pysal/access ***@***.***>
Cc: Julia Koschinsky ***@***.***>; Mention ***@***.***>
Subject: Re: [pysal/access] Partial solution to #18 (PR #26)


Great, can clean up once I have time to dig into the failing tests.

Yes, totally fine with switching to <=, though I think that's logically another issue.

@jkoschinsky<https://github.com/jkoschinsky> is great if she's willing!!

—
Reply to this email directly, view it on GitHub<https://github.com/pysal/access/pull/26#issuecomment-1030344957>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ADILY74RIGVQMN3CCSOYDCDUZQ45VANCNFSM5NQTIJWQ>.
Triage notifications on the go with GitHub Mobile for iOS<https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675> or Android<https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
You are receiving this because you were mentioned.Message ID: ***@***.***>

JamesSaxon commented:
Hi @knaaptime or @jGaboardi, please have a look if you like, and merge / close.

I've fixed the failing tests (it was a unit test failure¡), and also added a new class of tests, that assert:
* equality with a super-simple (20-line) implementation of the 2SFCA and 3SFCA
* that Eli's expectations of the relative access levels of different travel cost configurations.

Sorry it took me so long to get to this.

JamesSaxon commented:
Also tagging @jkoschinsky for info.

jGaboardi commented:
@knaaptime Can you give this PR a review?

JamesSaxon commented:
Hey @jGaboardi, waiting on @knaaptime, I just approved.

Anything else needed, for this to rebuild docs / version on the new site?

Or is that now all automagic?  Thanks again for all your support with this project.

@jkoschinsky for info.

jGaboardi commented:
@JamesSaxon sorry for the super late attention here. So we need a fresh release of `access`?

JamesSaxon commented:
Hi @jGaboardi, yeah -- the PR is in, so public version should be updated.

jGaboardi commented:
@JamesSaxon Are you thinking patch release (`v1.1.4`) or a feature release (`v1.2.0`)?

JamesSaxon commented:
Fair question -- patch release, in this case.
